connection_pool.go

```diff
@@ -49,23 +49,38 @@ func (b *backend) GetIdleConnection() *connection {
 // none are available then wait for up to the IdleConnectionWait time for one, else finally
 // establish a new connection if we aren't at the limit. If we are, then continue waiting
 // in increments of the idle time for a connection or the limit to come down before making
-// a new connection. This could potentially block up to the DialTimeout.
+// a new connection. This could potentially block up to twice the DialTimeout.
 //
 // If the error returned is NoConnectionsAvailable, the caller should treat it as transient
 // and not consider the backend/addr unhealthy.
 func (b *backend) GetConnection() (*connection, error) {
-	dialTimeout := time.After(b.conf.DialTimeout)
+	// dialTimeout must be longer than the configured timeout from the user to
+	// differentiate the case where 'the pool is full' and 'the remote server is
+	// not responding'. Since the b.conf.DialTimeout is used by the underlying
+	// newTCPConnection method, we need to still be alive and waiting if it returns
+	// an error at that point -- hence waiting for twice the configured timeout
+	// in this method.
```

Comment:
thanks for the explanation :)

Reply:
I'm having a "how did this ever work" feeling. Is the implication that previously, we were relying on the random jitter on IdleConnectionWait to determine whether we would try to build a new connection or not? Because if the entire random range is larger than DialTimeout, wouldn't we never create a new connection ever?